File: site/js/geometry.js
// Affine transforms and adaptive Bezier flattening.
// Mirrors clo2plt/geometry.py; the two must stay in step, since the Python
// side is the reference implementation the tests verify against.

export const IDENTITY = [1, 0, 0, 1, 0, 0];

export function multiply(m, n) {
  const [a, b, c, d, e, f] = m;
  const [A, B, C, D, E, F] = n;
  return [
    a * A + b * C,
    a * B + b * D,
    c * A + d * C,
    c * B + d * D,
    e * A + f * C + E,
    e * B + f * D + F,
  ];
}

export function apply(m, x, y) {
  return [m[0] * x + m[2] * y + m[4], m[1] * x + m[3] * y + m[5]];
}

export function scaleOf(m) {
  return (Math.hypot(m[0], m[1]) + Math.hypot(m[2], m[3])) / 2;
}

export function rotationOf(m) {
  return (Math.atan2(m[1], m[0]) * 180) / Math.PI;
}

// Python's round() uses banker's rounding; JS Math.round() does not. Plotter
// coordinates land on exact .5 often enough that this changes bytes, so the
// Python behaviour is reproduced here to keep outputs identical.
export function pyRound(value) {
  const floor = Math.floor(value);
  const diff = value - floor;
  if (diff > 0.5) return floor + 1;
  if (diff < 0.5) return floor;
  return floor % 2 === 0 ? floor : floor + 1;
}

export function flattenCubic(p0, p1, p2, p3, tol, out, depth = 0) {
  if (depth >= 24) {
    out.push(p3);
    return;
  }
  const [x0, y0] = p0;
  const [x3, y3] = p3;
  const dx = x3 - x0;
  const dy = y3 - y0;
  const chord = Math.hypot(dx, dy);

  if (chord < 1e-12) {
    const spread = Math.max(
      Math.hypot(p1[0] - x0, p1[1] - y0),
      Math.hypot(p2[0] - x0, p2[1] - y0)
    );
    if (spread <= tol) {
      out.push(p3);
      return;
    }
  } else {
    const d1 = Math.abs((p1[0] - x0) * dy - (p1[1] - y0) * dx) / chord;
    const d2 = Math.abs((p2[0] - x0) * dy - (p2[1] - y0) * dx) / chord;
    if (Math.max(d1, d2) <= tol) {
      out.push(p3);
      return;
    }
  }

  const p01 = [(x0 + p1[0]) / 2, (y0 + p1[1]) / 2];
  const p12 = [(p1[0] + p2[0]) / 2, (p1[1] + p2[1]) / 2];
  const p23 = [(p2[0] + x3) / 2, (p2[1] + y3) / 2];
  const p012 = [(p01[0] + p12[0]) / 2, (p01[1] + p12[1]) / 2];
  const p123 = [(p12[0] + p23[0]) / 2, (p12[1] + p23[1]) / 2];
  const mid = [(p012[0] + p123[0]) / 2, (p012[1] + p123[1]) / 2];

  flattenCubic(p0, p01, p012, mid, tol, out, depth + 1);
  flattenCubic(mid, p123, p23, p3, tol, out, depth + 1);
}

export function bbox(polylines) {
  let minx = Infinity, miny = Infinity, maxx = -Infinity, maxy = -Infinity;
  for (const pts of polylines) {
    for (const [x, y] of pts) {
      if (x < minx) minx = x;
      if (x > maxx) maxx = x;
      if (y < miny) miny = y;
      if (y > maxy) maxy = y;
    }
  }
  if (minx === Infinity) return null;
  return [minx, miny, maxx, maxy];
}

export function dedupe(points, eps = 1e-9) {
  const out = [];
  for (const p of points) {
    const last = out[out.length - 1];
    if (!last || Math.abs(p[0] - last[0]) > eps || Math.abs(p[1] - last[1]) > eps) {
      out.push(p);
    }
  }
  return out;
}

File: clo2plt/geometry.py
"""Affine transforms and adaptive Bezier flattening.

Flattening quality is the main lever on cut accuracy: the plotter reproduces
exactly the polyline we hand it, so the chord tolerance here becomes real
deviation on fabric. Subdivision is recursive with a flatness test, which bounds
the error, rather than a fixed step count, which does not.
"""

import math

# PDF matrix [a b c d e f] == [[a b 0],[c d 0],[e f 1]]
IDENTITY = (1.0, 0.0, 0.0, 1.0, 0.0, 0.0)


def multiply(m, n):
    """Return m concatenated with n (apply m first, then n)."""
    a, b, c, d, e, f = m
    A, B, C, D, E, F = n
    return (
        a * A + b * C,
        a * B + b * D,
        c * A + d * C,
        c * B + d * D,
        e * A + f * C + E,
        e * B + f * D + F,
    )


def apply(m, x, y):
    a, b, c, d, e, f = m
    return (a * x + c * y + e, b * x + d * y + f)


def scale_of(m):
    """Mean linear scale of a matrix, for converting tolerances across spaces."""
    a, b, c, d, _, _ = m
    return (math.hypot(a, b) + math.hypot(c, d)) / 2.0


def rotation_of(m):
    """Rotation angle in degrees implied by a matrix's x-axis."""
    a, b = m[0], m[1]
    return math.degrees(math.atan2(b, a))


def flatten_cubic(p0, p1, p2, p3, tol, out, depth=0):
    """Append points after p0 approximating the cubic within `tol`.

    Flatness test measures control-point deviation from the p0->p3 chord, which
    upper-bounds the true curve deviation.
    """
    if depth >= 24:
        out.append(p3)
        return

    x0, y0 = p0
    x3, y3 = p3
    dx, dy = x3 - x0, y3 - y0
    chord = math.hypot(dx, dy)

    if chord < 1e-12:
        # Degenerate chord (closed loop): fall back to control-hull extent so a
        # curve that returns to its origin still gets subdivided rather than
        # collapsing to a point.
        spread = max(
            math.hypot(p1[0] - x0, p1[1] - y0), math.hypot(p2[0] - x0, p2[1] - y0)
        )
        if spread <= tol:
            out.append(p3)
            return
    else:
        d1 = abs((p1[0] - x0) * dy - (p1[1] - y0) * dx) / chord
        d2 = abs((p2[0] - x0) * dy - (p2[1] - y0) * dx) / chord
        if max(d1, d2) <= tol:
            out.append(p3)
            return

    # de Casteljau split at t=0.5
    p01 = ((x0 + p1[0]) / 2, (y0 + p1[1]) / 2)
    p12 = ((p1[0] + p2[0]) / 2, (p1[1] + p2[1]) / 2)
    p23 = ((p2[0] + x3) / 2, (p2[1] + y3) / 2)
    p012 = ((p01[0] + p12[0]) / 2, (p01[1] + p12[1]) / 2)
    p123 = ((p12[0] + p23[0]) / 2, (p12[1] + p23[1]) / 2)
    mid = ((p012[0] + p123[0]) / 2, (p012[1] + p123[1]) / 2)

    flatten_cubic(p0, p01, p012, mid, tol, out, depth + 1)
    flatten_cubic(mid, p123, p23, p3, tol, out, depth + 1)


def bbox(polylines):
    """Bounding box (minx, miny, maxx, maxy) over an iterable of point lists."""
    minx = miny = math.inf
    maxx = maxy = -math.inf
    for pts in polylines:
        for x, y in pts:
            if x < minx:
                minx = x
            if x > maxx:
                maxx = x
            if y < miny:
                miny = y
            if y > maxy:
                maxy = y
    if minx is math.inf:
        return None
    return (minx, miny, maxx, maxy)


def dedupe(points, eps=1e-9):
    """Drop consecutive duplicate points, which plotters render as pen jitter."""
    out = []
    for p in points:
        if not out or abs(p[0] - out[-1][0]) > eps or abs(p[1] - out[-1][1]) > eps:
            out.append(p)
    return out
